### ISSUE TITLE
Android - Some devices unable to get product info, unable to place orders.

### ISSUE THREAD

**Issue by EEETA** [j3k0/cordova-plugin-purchase#1613]
### Observed behavior

Hi, I'm migrating my plugin from v11.x to v13.11.1 due to Google Play requirements (upgrading Play Billing from v4 to v7). I modified my old code based on the [migration doc](https://github.com/j3k0/cordova-plugin-purchase/wiki/HOWTO:-Migrate-to-v13) after upgrading the plugin library.

For iPhone, product info can be obtained normally. **However, for Android devices, some are able to get the info while others cannot.**  

```
Project environment：

> node: v20.14.0
> angular: "1.7.9",
> cordova-ios: "^5.1.1",
> cordova-android: "^12.0.1",
> cordova-plugin-purchase: "^13.11.1",
```

### Devices info and issues:
In China, phones cannot use Google services without rooting. But rooting may also remove some features and not be as full-featured as the original system.
- [ ] <span style="color:red">2 of 4 Android devices can get info normally but **Redmi k50** gets undefined (Redmi Note9 can't use Google Play so ignore)  </span>
- [ ] **Redmi k50** can't get product from Play so can't place order in Android or do sandbox pay, but worked in v11.  

![image](https://github.com/user-attachments/assets/ec7e2154-af4a-4e28-b3ef-3b7be5a3fade)

![image](https://github.com/user-attachments/assets/3298e0b8-a0dc-47c4-995a-945240e6b4c8)

I confirmed using same apk and product IDs. I don't understand why on the same phone, the old version using the plugin v11 could get product and support placing orders, while the migrated new version can't get product info. If it's about product registration failure, then why the other two Android phones can get products normally?

### Code
![Pasted Graphic 4](https://github.com/user-attachments/assets/e35f0e3e-f037-419b-818b-aefa99ca0765)
![image](https://github.com/user-attachments/assets/9f591f6a-e634-4a85-8d0d-5b6595eeb75b)
![image](https://github.com/user-attachments/assets/618e2210-1872-42cf-9d13-392b08c66bae)
![image](https://github.com/user-attachments/assets/d6e997a9-bfd4-42e5-ad35-4b4e400639c0)

This issue has puzzled me for days. Please offer your help and answer my questions.Thanks.

**Comment by john290n** [comment]
Hi @EEETA, Im facing the same issue on android since Im using the cordova-plugin-purchase v13.11.1 only on android because for iOS it worked well. The error that Im getting when I try to register products is “Product with this id not found” and it is very weird. The products are undefined everytime. I did all needed steps:
1. By creating products, subscriptions on the Google Play console
2. My android app is deployed on production, I pushed a new release to closed and internal testing
3. I logged in with the license user on my android physical device

I use React Js and capacitor.

**Comment by john290n** [comment]
I did research for many days, but I didn’t find anything to fix it. 😕 
Please let me know if you have fixed it

**Comment by EEETA** [comment]
@john290n 
You can try the following steps one by one (best to prepare multiple Android devices):

1. Make sure you have correctly uploaded the products to the store and successfully configured the products (the configuration may not take effect in real-time, best to configure in advance), and check their activation status.
2. If you need to use sandbox simulated payment, the test users also need to be configured in advance for 24 hours, and it may take effect the next day. Log in to the Android test device you prepared.
3. Make sure you have filled in any information required by Google Play (such as country/region, app details, payment information, etc.) and sent it for review.
4. Check the versionCode of the testing package installed locally, which needs to be consistent with the version on the Google Play track.
5. Before actually getting your custom product information, you can try connecting to the [Platform.TEST](https://github.com/j3k0/cordova-plugin-purchase/blob/c5cb3db99f02d8a908e67bae4658db1f30bc24ff/api/modules/CdvPurchase.Test.md?plain=1#L30) in the plugin first and try to get the TEST product from the store.
```ruby
CdvPurchase.store.register([{
        type: CdvPurchase.ProductType.CONSUMABLE,
        id: 'test-consumable',
        platform: CdvPurchase.Platform.TEST,
}]);

const product = CdvPurchase.store.get('test-consumable, CdvPurchase.Platform.TEST);
alert(JSON.stringify(product))
```
&nbsp;
***Following the first 5 steps, it should prove that the registration logic of your code is correct (getting Platform.TEST and PlatformAPPLE_APPSTORE normally).***

You can try the final step,

```
6. getting the product from Platform.GOOGLE_PLAY.

6.1 Make sure the plugin is loaded normally after the Android app starts.

6.2 Print out CdvPurchase, CdvPurchase.store and your ProductId before store.register (make sure the log content exists and the id is consistent with the configuration on Google Play).

6.3 Execute CdvPurchase.store.initialize([..., CdvPurchase.Platform.GOOGLE_PLAY]) at the end of initialization.

6.4 You can manually store.update() to refresh the current product prices and status before executing store.get().

6.5 Make sure the id and platform parameters in store.get(id, platform) are correct.

6.6 If no problems, you should at least be able to get the product information by this point. (If still unable to retrieve, try Step 6 again with a different Android device.)
```

As for my Redmi k50 being unable to retrieve products, I currently suspect it may be due to an update to Google services or require further iteration of the plugin. I still have not found a solution and am waiting for answers.

**Comment by john290n** [comment]
Many thanks @EEETA, I will try to do all steps above!

**Comment by john290n** [comment]
@EEETA I tried your steps but it is not working.

These are the logs that Im getting on Android Studio:
> I/Capacitor/Console: File: http://localhost/static/js/8.00a9bea1.chunk.js - Line 1 - Msg: Create CdvPurchase...
I/Capacitor/Console: File: http://localhost/static/js/45.62950d65.chunk.js - Line 1 - Msg: LENGTH OF PRODUCTS 
I/Capacitor/Console: File: http://localhost/static/js/45.62950d65.chunk.js - Line 1 - Msg: Product Loaded: undefined
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase] INFO: initialize(["android-playstore","paid subscription"]) v13.11.1
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: Adding platforms: [{"platform":"android-playstore"},{"platform":"paid subscription"}]
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: 
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: 
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: GooglePlay initializing...
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] INFO: Initialize
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay.Bridge] INFO: setup ok
I/Capacitor/Console: File: http://localhost/static/js/45.62950d65.chunk.js - Line 1 - Msg: LENGTH OF PRODUCTS 
V/Capacitor/Plugin: To native (Cordova plugin): callbackId: InAppBillingPlugin78179932, service: InAppBillingPlugin, action: setListener, actionArgs: []
V/Capacitor/Plugin: To native (Cordova plugin): callbackId: InAppBillingPlugin78179933, service: InAppBillingPlugin, action: init, actionArgs: []
D/CdvPurchase: sendToListener() -> ready
D/CdvPurchase:             data -> {}
D/CdvPurchase: init()
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay.Bridge] INFO: listener: {"type":"ready","data":{}}
D/CdvPurchase: startServiceConnection()
I/Capacitor/Console: File: http://localhost/static/js/45.62950d65.chunk.js - Line 1 - Msg: LENGTH OF PRODUCTS 
D/CdvPurchase: startServiceConnection() -> Success
D/CdvPurchase: init() -> Success
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: Ready
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: GooglePlay initialized. 
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: GooglePlay products: [{"type":"paid subscription","id":"1_month_premium","platform":"android-playstore"}]
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: Load: [{"type":"paid subscription","id":"1_month_premium","platform":"android-playstore"}]
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: getAvailableProducts: [] | ["1_month_premium"]
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay.Bridge] INFO: getAvailableProducts()
V/Capacitor/Plugin: To native (Cordova plugin): callbackId: InAppBillingPlugin78179934, service: InAppBillingPlugin, action: getAvailableProducts, actionArgs: [[],["1_month_premium"]]
D/CdvPurchase: getAvailableProducts()
D/CdvPurchase: queryAllProductDetails()
D/CdvPurchase: queryAllProductDetails() -> Query SUBS.
D/CdvPurchase: queryProductDetailsAsync()
D/CdvPurchase: executeServiceRequest() -> OK
D/CdvPurchase: queryProductDetailsAsync() -> Success
W/CdvPurchase: queryAllProductDetails() -> Query returned nothing.
D/CdvPurchase: queryAllProductDetails() -> Calling listener.
D/CdvPurchase: getAvailableProducts() -> Success
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: Loaded: []
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: getPurchases
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay.Bridge] INFO: getPurchases()
V/Capacitor/Plugin: To native (Cordova plugin): callbackId: InAppBillingPlugin78179935, service: InAppBillingPlugin, action: getPurchases, actionArgs: ["null"]
D/CdvPurchase: getPurchases()
D/CdvPurchase: queryPurchases()
D/CdvPurchase: executeServiceRequest() -> OK
I/CdvPurchase: queryPurchases(SUBS) -> Elapsed time: 29ms
I/CdvPurchase: queryPurchases(INAPP) -> Elapsed time: 37ms
D/CdvPurchase: sendToListener() -> setPurchases
D/CdvPurchase:             data -> {"purchases":[]}
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay.Bridge] INFO: listener: {"type":"setPurchases","data":{"purchases":[]}}
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: onSetPurchases: []
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: onPurchaseUpdated: 
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.AdapterListener] DEBUG: receiptsReady: android-playstore (1/0)
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.GooglePlay] DEBUG: getPurchases success
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: GooglePlay products loaded: [{"isError":true,"code":6777012,"message":"Product with id 1_month_premium not found.","platform":"android-playstore","productId":"1_month_premium"}]
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.Adapters] INFO: GooglePlay receipts loaded: []
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.AdapterListener] DEBUG: setSupportedPlatforms: android-playstore (1 have their receipts ready)
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.AdapterListener] DEBUG: triggering receiptsReady()
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase] DEBUG: Calling callback: type=receiptsReady() name=receiptsMonitor_setup reason=adapterListener_setSupportedPlatforms
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.ReceiptsMonitor] DEBUG: receiptsReady...
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.ReceiptsMonitor] DEBUG: check(0/0)
I/Capacitor/Console: File: http://localhost/ - Line 3209 - Msg: [CdvPurchase.ReceiptsMonitor] INFO: receiptsVerified()

The subscription seems to be active in Google Play Console
<img width="1146" alt="Screenshot 2024-10-19 at 10 54 17" src="https://github.com/user-attachments/assets/e762c88a-259f-4041-9eef-6ddced9a01d4">


And here is my code in React Js:

import React, {useState, useEffect, useContext} from 'react'
import 'cordova-plugin-purchase/www/store'
import Badge from '@material-ui/core/Badge'
import routes from 'routes'
import {useHistory, generatePath} from 'react-router-dom'
import LoadingCircle from 'components/LoadingCircle'
import I18nContext from 'contexts/I18nContext'
import {
  Wrapper,
  ButtonTypes,
  PartWrapper,
  Top,
  FavouritePlanWrapper,
  TopIcon,
  Number,
  String,
} from './AndroidInAppPurchase.styles'

const defaultProps = {
  color: 'primary',
}

const AndroidInAppPurchase = () => {
  const {t} = useContext(I18nContext)
  const [storeState, setStoreState] = useState([])
  const history = useHistory()
  const {store, ProductType, Platform} = window.CdvPurchase

  useEffect(() => {
    window.CdvPurchase.store.verbosity = window.CdvPurchase.LogLevel.DEBUG

    const initializeStore = async () => {
      try {
        store.register([
          {
            type: ProductType.PAID_SUBSCRIPTION,
            id: '1_month_premium',
            platform: Platform.GOOGLE_PLAY,
          },
        ])

        const product = store.get('1_month_premium', Platform.GOOGLE_PLAY)
        console.log('Product Loaded:', product)
        setStoreState([product])
        store.initialize([Platform.GOOGLE_PLAY, ProductType.PAID_SUBSCRIPTION])
      } catch (error) {
        console.error('Store initialization failed:', error)
      }
    }

    initializeStore()
  }, [])

  console.log('LENGTH OF PRODUCTS', storeState)

  return (
    <Wrapper>
      {storeState.length === 0 ? (
        <LoadingCircle isAbsolute />
      ) : (
        <>
          {storeState?.map(product => (
            <>
              <PartWrapper>
                <Top>
                  {product?.raw?.billingPeriod === 1 &&
                    product?.raw?.billingPeriodUnit === 'Month' && (
                      <FavouritePlanWrapper>
                        <Badge badgeContent={t('getPremium.favourite')} {...defaultProps} />
                      </FavouritePlanWrapper>
                    )}
                  <TopIcon>
                    <Number>
                      {product?.raw?.billingPeriodUnit === 'Year'
                        ? 12
                        : product?.raw?.billingPeriod}
                    </Number>
                    <String>
                      {product?.raw?.billingPeriod === 1 &&
                      product?.raw?.billingPeriodUnit === 'Month'
                        ? 'month'
                        : 'months'}
                    </String>
                  </TopIcon>
                </Top>
              </PartWrapper>
              <ButtonTypes
                color='secondary'
                onClick={() =>
                  history.push(
                    generatePath(routes.getPremium.billingAddress.path, {id: product.id})
                  )
                }
              >
                {t('getPremium.iosPrice')} {product?.raw?.price}
              </ButtonTypes>
              <br />
            </>
          ))}
        </>
      )}
    </Wrapper>
  )
}

export default AndroidInAppPurchase

**Comment by john290n** [comment]
Im testing it locally with Android physical device, by connecting it with Wifi with Android studio

**Comment by EEETA** [comment]
@john290n 
I saw from your logs that the product cannot be loaded with the message "Line 1 - Msg: Product Loaded: undefined". Did you run the tests from step 5? You can first connect to Platform.TEST to see, if step 5 can retrieve the TEST product. Then you can try using other physical Android devices. Currently I don't have any other suggestions, I still think there may be compatibility issues between the phone itself, the plugin and Google Play.

**Comment by john290n** [comment]
@EEETA I tried with TEST on web browser by running the React Js on web and Im getting these results:
<img width="639" alt="Screenshot 2024-10-19 at 11 45 17" src="https://github.com/user-attachments/assets/fbaa9c70-c0ab-42d6-ae8a-bc684f06a534">

**Comment by EEETA** [comment]
@john290n 
In Platform.TEST, the id and platform need to correspond one-to-one, otherwise it will show prompts such as product unavailable if set incorrectly. Among them, CdvPurchase.ProductType.CONSUMABLE type corresponds to the "test-consumable" id. If you want to test other types, you should click the link to the fifth step I provided to find the type you need.
![image](https://github.com/user-attachments/assets/11357321-362b-40ef-b1c1-b263822cac20)

**Comment by john290n** [comment]
@EEETA Yes, you're right—it worked on the web. However, how can I be sure that in-app purchases with subscriptions will work in production? When I tried using the product IDs I set up in the Google Play Console with the platform set to GOOGLE_PLAY, it wasn’t working.


<img width="636" alt="Screenshot 2024-10-19 at 11 59 05" src="https://github.com/user-attachments/assets/edc23d5a-609c-450c-8252-2ad427aefce5">

**Comment by EEETA** [comment]
@john290n 
Yes, after validation, we know that the registration logic of your code itself is normal, and your product type is also supported. For situations where Google Play cannot be retrieved, when I tried switching to other Android devices, it was able to on other Android devices, other possibilities may only need to wait for the plugin author to adjust.

**Comment by john290n** [comment]
@EEETA  So it means that my code is okay? Now if I want to try it out in Android physical device with Android studio locally I can just change the Platform.GOOGLE_PLAY, and the id of product?

**Comment by john290n** [comment]
@EEETA  Could you please provide me any screenshot of the steps that you did to setup the Products / Subscriptions on your Google Play?

**Comment by EEETA** [comment]
@john290n 
Currently it seems that your code is correct, the bigger reason is the problem between the phone itself, the plugin and Google Play, because you can correctly retrieve information from Platform.TEST and Platform.APPLE_APPSTORE, only GooglePlay is not working. However, you can still try one more approach - try each type one by one, when I tried before it was unrelated to type, if you have other Android devices, it is recommended to switch to other test devices for testing.

![image](https://github.com/user-attachments/assets/3d4dd125-1d59-41ab-8ea2-50b5cfaf335d)
My product type is different from yours, I have set it as CONSUMABLE.

**Comment by john290n** [comment]
@EEETA Aha, so you created the product in the In-app products section and not under the Subscription tab. How do I use the license testing user when testing locally? Currently, I'm just logged in with this user on my Samsung Galaxy account.

**Comment by EEETA** [comment]
@john290n 
Yes, I am in the sandbox environment, which can help me do store payments in the local test environment, but extra configuration is required, and the specific configuration steps are also more complicated. You can try looking up relevant documents for guidance.

**Comment by john290n** [comment]
@EEETA, have you checked the logs? Are your logs similar to mine when the product with this ID is not found? I just want to make sure we’re experiencing the same issue.

**Comment by EEETA** [comment]
@john290n It's the same issue, and the logs are similar as well, but I am using web chrome debugging, the only difference is that our product IDs and types are different.

**Comment by john290n** [comment]
@EEETA, do you have any idea if this issue is going to persist? Are you considering switching to a different library?

**Comment by EEETA** [comment]
@john290n I'm not sure when this issue will be resolved. I'm trying to communicate with the plugin author but haven't received a response yet. I will continue waiting.

**Comment by john290n** [comment]
@EEETA how can we verify that the products in the Google Play Console are OK and valid to use?

**Comment by EEETA** [comment]
@john290n It should be that after store.register, if you can get the product information and the status is normal, then it indicates it's available. But in the documentation for plugin v13, I didn't see any relevant description for the register part, I feel the documentation description is somewhat lacking. I confirmed this by setting the correct and incorrect ids. If the id is different from GooglePlay, it will be unable to get the product or prompt other errors. As for whether to verify the product is available, I can only initiate store.order(), there may be other methods provided by the plugin but I'm not very clear, this is also what I'm currently trying to figure out.

**Comment by john290n** [comment]
@EEETA Yes, but I just set up the merchant account in the Monetize tab, and I created the in-app products. I’m not sure if there are any other steps needed. My app is in production, and all other steps are complete.

<img width="268" alt="Screenshot 2024-10-19 at 13 03 19" src="https://github.com/user-attachments/assets/e35d15f1-b7c7-426d-ab95-bbcb4761fab6">

**Comment by john290n** [comment]
@EEETA, I just noticed that my payment profile on the Google Play Console is not set up yet. Could this be the issue with the undefined products?

**Comment by EEETA** [comment]
@john290n Sorry, I'm not certain if you definitely need to fill in this information. But based on my search results and viewing previous issues, I found that many people said it's necessary to fill in the GooglePlay & AppStore information (including the original author) properly. So it's best if you configure it as well.

**Comment by john290n** [comment]
@EEETA thank you, please let me know if you find something to fix the issue!

**Comment by john290n** [comment]
@EEETA did you find any solution?

**Comment by Dexus** [comment]
You have to fill the Company Informations, the Financial Informations and have the App Version published as alpha/beta to the Stores. Then for development, you should keep the version that is released to alpha/beta for local development and testing.

Once your Profile on Google and Apple Side are complete filled and accepted, you can "test" your in app purchases. Not before your have don it completely. And AGAIN: Fill all Documents that are required by Google and Apple - if not your will never be able to test your IAP.

**Comment by john290n** [comment]
Hi @Dexus, I have completed the Payment Profile in Google Play console but Im still getting the same error 😕 (Product with id "1_month_premium" not found")

**Comment by Dexus** [comment]
Only on a specific phone or at all?
Did you use the correct app identifer? com.company.app?

Build system Cordova? PhoneGap? Ionic / Capcitorjs?

**Comment by john290n** [comment]
It is happening at all phones, yes Im using cordova-plugin-purchase v13.11.1, and I think that my app identifier is correct!

**Comment by john290n** [comment]
@Dexus, are you available for a quick meeting? Your help would be greatly appreciated. 🙏

**Comment by john290n** [comment]
I’ve been trying to fix it for more than two weeks without success.

**Comment by EEETA** [comment]
@john290n Sorry, I've been busy with tasks these past few days. I have already completed the entire purchase process and made it compatible with Android and iOS phones, preparing to launch the product. Here are some thoughts that may help you:

1. Recheck the Google Play configuration to ensure the app settings and entire account settings are correct. (Pictures will be attached below)
2. If the above settings are correct, check that the Android device you are currently logged into only has test accounts logged in. If there are multiple accounts, you need to log out of those not on the test list.
3. It's best if possible to test using different physical devices, you can summarize the information of the Android device you used and send it back to me.
4. Regarding the situation where my Redmi K50 cannot use Google Play, when I output the error, the error text contains "FEATURE_NOT_SUPPORTED", error code is ErrorCode.LOAD, this is only briefly mentioned in the plugin documentation. I suggest separately prompting the user that the current device does not support it (also because Google Play itself does not support phones within mainland China, and previous older versions may have been able to use without strict detection by Google Play or plugin, while the new version added detection. Of course, this is just my guess, because we tested over a dozen phones, and only two mainland Chinese phones had abnormalities, while others were normal.)
5. There is also an additional situation where the store supports it but the product fails to register due to slow network, then a retry is needed, so you need to tell the user to please retry and ensure your code performs registration at the deviceready stage.
6. You can try obtaining the full message in store.error to see if the plugin provides other messages.

![image](https://github.com/user-attachments/assets/d53bb4c7-d102-4d67-960d-14d3f42b9d36)
![image](https://github.com/user-attachments/assets/3aa5de0e-d8cf-44dc-9418-2b2ab2059593)
![image](https://github.com/user-attachments/assets/61d4972f-2e04-4d2c-b5f8-c981da334790)
![image](https://github.com/user-attachments/assets/07421151-5bb6-4788-94e1-4c874efe44b2)
